### PR TITLE
admin edit user accounts

### PR DESCRIPTION
**Pull request by DeveloperX104**

### CODE REVIEW

**Review by 2kama — CHANGES_REQUESTED** [review]
I sent a video to your whatsapp with a bunch of changes that need be made

File: src/components/Admin/SelectedUserModal.js [hunk 1]
```diff
@@ -114,11 +124,9 @@ const SelectedUserModal = ({showModal, closeModal, removeEntry=false, selectedAl
                 {selectedAlterUser.role && selectedAlterUser.role.isAffiliate && 
                     <FormField name="affiliateCode" placeholder="Affiliate Code" type="text" />
                 }
-
-                {!removeEntry && <FormField name="referralCode" placeholder="Referral Code" type="text" />}
```
Comment (2kama) [inline src/components/Admin/SelectedUserModal.js:118]: Why was this deleted?

And when put back, it also only shows when the account isUser

File: src/pages/AdminPages/AdminDashboard/index.jsx [hunk 2]
```diff
@@ -1,32 +1,63 @@
...
+                        <button disabled={alterUsers[0] && alterUsers[0].role.isAffiliate  && true} onClick={() => {update(false, getAffiliates)}}>View Affiliate Accounts</button> <br />
+                        <button disabled={alterUsers[0] && alterUsers[0].role.isTrader  && true} onClick={() => {setRemoveEntry(false); dispatch(setAlterUsers(traders))}}>View Traders Accounts</button> <br />
+                        <button disabled={alterUsers[0] && alterUsers[0].role.isAdmin  && true} onClick={() => {update(false, getAdmins);}}>View Admin Accounts</button> <br />
 
+                        <hr />
+                        {
+                        isLoading ? 
+                        <>Loading...</>:
+                            alterUsers.length > 0 ? <AlterUsers users={alterUsers} approve={removeEntry} /> : <>No User Found For This Category</>
+                        }
```
Comment (2kama) [inline src/pages/AdminPages/AdminDashboard/index.jsx]: If i were to guess, you are now loading the table of users on the dashboard Page... that's a big no!

We can't deteremine what will be on dashboard now.... It will be preferable to have a separate page to show all the different kind of users.

If we are then going with the route of having all users showcase on one page, there'll be no use of putting AlterUsers and SelectedUserModal components in the Component folder since they'll only be used by this one page.

**Review by 2kama — DISMISSED** [review]

File: src/components/Admin/SelectedUserModal.js (hunk 1 above)
Comment (DeveloperX104) [reply src/components/Admin/SelectedUserModal.js:118]: I thought about it again, `referralCode` should only be set when registering, there's no need to be able to set that field after the user has registers.

File: src/pages/AdminPages/AdminDashboard/index.jsx (hunk 2 above)
Comment (DeveloperX104) [reply src/pages/AdminPages/AdminDashboard/index.jsx]: Well, we can have a couple of data on the dashboard page. So having this data on the dashboard page isn't bad, concerning the file location I can move the folder from the `components` folder and place inside the `Admin` components in `pages`.

File: src/components/Admin/SelectedUserModal.js (hunk 1 above)
Comment (2kama) [reply src/components/Admin/SelectedUserModal.js:118]: The admin wants to be able to edit it. So on edit users, it's a field that the admin can alter.

File: src/pages/AdminPages/AdminDashboard/index.jsx (hunk 2 above)
Comment (2kama) [reply src/pages/AdminPages/AdminDashboard/index.jsx]: At this point, yes, you are right... but lay no focus on the dashboard, we can't know what should be there until we are done with the app.

Recreate 2 new pages (the ones you deleted)... one for approving accounts and another for edit accounts... there's a whole lot more that can be viewed in the edit accounts page...it's literally going to be an information page for viewing user's activity logs, transaction logs, active portfolios etc.... approving page is just for approving.

I don't think there was anything wrong with the approval page (please add back). This branch is for creating an edit account page (please focus on that).

Comment (DeveloperX104) [reply src/pages/AdminPages/AdminDashboard/index.jsx]: Alright, thanks.
Can you suggest a nice URL for these two?
I'm thinking instead of haaving 2 files for these 2 things, I'd just use 1 and determine what the admin wants to request for from the URL then use that to fetch the right data

File: src/components/Admin/SelectedUserModal.js [hunk 3]
```diff
@@ -124,7 +126,7 @@ const SelectedUserModal = ({showModal, closeModal, removeEntry=false, selectedAl
                 {selectedAlterUser.role && selectedAlterUser.role.isAffiliate && 
                     <FormField name="affiliateCode" placeholder="Affiliate Code" type="text" />
                 }
-             
+                <FormField name="referralCode" placeholder="Referral Code" type="text" />
```
Comment (2kama) [inline src/components/Admin/SelectedUserModal.js]: `{!removeEntry && <FormField name="referralCode" placeholder="Referral Code" type="text" disabled />}`

File: src/pages/AdminPages/ApprovedAccounts/index.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    return(
+        <>
+            <Authenticate inside={true} />
+            {
+                role && (
+                    role.isAdmin ? (
+                    <>
+                        <button disabled={alterUsers[0] && alterUsers[0].role.isUser && !alterUsers[0].role.isAffiliate  && true} onClick={() => {update(getUsers)}}>View User Accounts</button> <br />
```
Comment (2kama) [inline src/pages/AdminPages/ApprovedAccounts/index.js]: Then what happens if the first entry in the array is an affiliate even though we are trying to get users who can also be affiliates?

File: src/pages/AdminPages/ApprovedAccounts/index.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                        <button disabled={alterUsers[0] && alterUsers[0].role.isAffiliate  && true} onClick={() => {update(getAffiliates)}}>View Affiliate Accounts</button> <br />
```
Comment (2kama) [inline src/pages/AdminPages/ApprovedAccounts/index.js]: And what if the first entry in the array is an affiliate which is also in the user list?

I guess my point is (and it's not really important now), disabling the buttons in not neccessary... but we can deal with that later.

File: src/pages/AdminPages/ApprovedAccounts/index.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                        <button disabled={alterUsers[0] && alterUsers[0].role.isTrader  && true} onClick={() => {dispatch(setAlterUsers(traders))}}>View Traders Accounts</button> <br />
+                        <button disabled={alterUsers[0] && alterUsers[0].role.isAdmin  && true} onClick={() => {update(getAdmins);}}>View Admin Accounts</button> <br />
+                        <hr />
+                        {!isLoading ? alterUsers.length > 0 ? <AlterUsers users={alterUsers} approve={false} /> : <>No Approved Accounts For this category</>: <>Loading...</>}
```
Comment (2kama) [inline src/pages/AdminPages/ApprovedAccounts/index.js]: Yeah, I noticed you are actually toggling `isLoading`... `isLoading` is built just for getting user data. You have to device another means of checking if other data is steal being fetch... I'll suggest setting their default to `undefined` or `null`, then checking on those to know when your apis are down fetching...

File: src/pages/AdminPages/UnapprovedAccounts/index.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    return(
+        <>
+            <Authenticate inside={true} />
+            {
+                !isLoading && role && (
+                role.isAdmin ? (
+                    <>
+                        Unapproved Users
+                        <hr />
+                        {!isLoading ? alterUsers.length > 0 ? <AlterUsers users={alterUsers} approve={true} /> : <>No Unapproved Accounts</>: <>Loading...</>}
```
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/index.js]: `isLoading` is for userData...

File: src/pages/AdminPages/ApprovedAccounts/index.js (hunk 6 above)
Comment (DeveloperX104) [reply src/pages/AdminPages/ApprovedAccounts/index.js]: Ooh thanks, how about I create another `isLoading` stae for fetching just `adminQuery`, I think that's better.

File: src/pages/AdminPages/ApprovedAccounts/index.js (hunk 5 above)
Comment (DeveloperX104) [reply src/pages/AdminPages/ApprovedAccounts/index.js]: Well, disabling the button isn't important, but we need to be able to get which button is clicked last so as to know the kind of data that was rendered.

File: src/pages/AdminPages/ApprovedAccounts/index.js (hunk 4 above)
Comment (DeveloperX104) [reply src/pages/AdminPages/ApprovedAccounts/index.js:40]: One of the main reason why I decided to disable the button is to know which button was clicked and which data is being rendered, but now I'd just create a new column for the roles so the admin knows which data is being rendered.

File: src/pages/AdminPages/ApprovedAccounts/index.js (hunk 5 above)
Comment (DeveloperX104) [reply src/pages/AdminPages/ApprovedAccounts/index.js]: So what I did instead os to have a `ROLE` field to let the admin know which category of data is being rendered in the `Table` component.

File: src/components/Admin/AlterUsers.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    const getRole = role => {
+        if (role.isAffiliate) return 'AFFILIATE'
+        if (role.isUser) return 'USER'
+        if (role.isTrader) return 'TRADER'
+        if (role.isSuperAdmin) return 'SUPER ADMIN'
+        if (role.isAdmin) return 'ADMIN'
+    }
+
+    const disableButton = user => (user.role.isSuperAdmin && selectedAlterUser.role && !selectedAlterUser.role.isSuperAdmin) || user.uid === currentUser.uid
```
Comment (2kama) [inline src/components/Admin/AlterUsers.js]: The first condition here is wrong. SuperAdmins can edit other SuperAdmins. Imagine if the client did have another superAdmin and later on wanted to demote him/her back to a normal Admin.

And I don't get why you're checking `selectedAlterUser.role`, `selectedAlterUser` will not be populated until a button is clicked...so checking it before a click of a button doesn't gave any data whatsoever.

so this should be : 

`const disableButton = user => (user.role.isSuperAdmin && !currentUser.role.isSuperAdmin) || user.uid === currentUser.uid`

File: src/components/Admin/AlterUsers.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            dispatch(deleteUnapprovedUser(user))
+            dispatch(setAlterUsers(
+                users.filter(u => u.uid !== user.uid)
+            ))
+        }
+    }
+
+    const getRole = role => {
+        if (role.isAffiliate) return 'AFFILIATE'
+        if (role.isUser) return 'USER'
```
Comment (2kama) [inline src/components/Admin/AlterUsers.js]: not a blocker ... change this to `'NORMAL USER'`

File: src/components/Admin/AlterUsers.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+    const openModalAndAlterUser = (user)  => {
+        dispatch(setSelectedAlterUser({...user}))
+        setShowModal(true)
+    }
+
+    const deleteUser = (user) => {
+        if (window.confirm("Are you sure you want to delete this account?")) {
+            dispatch(disableButton())
+            dispatch(deleteUnapprovedUser(user))
```
Comment (2kama) [inline src/components/Admin/AlterUsers.js]: Since we are gonna be using this page as a component for both `Approved` and `Unapproved` users. It'll be best to rename this function with something more generic....maybe `deleteUserData`.

Please find all instances and connection of this dispatch function in the `reducers`, `constants`, `sagaRoot`, `handler` and `request` files and change accordingly as well

File: src/pages/AdminPages/AdminDashboard/index.jsx [hunk 11]
```diff
@@ -1,32 +1,43 @@
...
-        isLoading : state.isLoading
-    }))
+        isLoading : state.user.isLoading,
+        alterUsers: state.adminQuery.alterUsers,
+        traders: state.adminQuery.traders,
+    }), shallowEqual)
+
+    useEffect(() => {
+        dispatch(setIsLoading());
+        dispatch(getTraders())
```
Comment (2kama) [inline src/pages/AdminPages/AdminDashboard/index.jsx]: move this dispatch call into the `AlterUsers` file. by doing this here you've finalized that we must have the traders array persist into any page we link from here... and remember I said, we cannot make any assumptions for what will happen on the dashboards for now... not until we are done with the inner functions.

File: src/pages/AdminPages/ApprovedAccounts/index.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        role : state.user.role,
+        isLoading : state.adminQuery.isLoading,
+        alterUsers: state.adminQuery.alterUsers,
+        traders: state.adminQuery.traders,
+    }), shallowEqual)
+
+    useEffect(() => {
+        dispatch(setIsLoading());
+        dispatch(getUsers())
+        dispatch(getTraders())
```
Comment (2kama) [inline src/pages/AdminPages/ApprovedAccounts/index.js:23]: If you are getting traders here in the `Approved` page (which I think is a good idea), send the traderArray as props into the `AlterUsers` component. Do same for `Unapproved` page.

**Review by 2kama — APPROVED** [review]
